rmw_fastrtps_cpp/src/register_node.cpp

```diff
@@ -0,0 +1,217 @@
+        eprosima_fastrtps_identifier,
+        participant_info.get(),
+        pub))
+      {
+        fprintf(
+          stderr, "Failed to destroy publisher after function: '"
+          RCUTILS_STRINGIFY(__function__) "' failed.\n");
+      }
+    });
+  if (!publisher) {
+    return RMW_RET_BAD_ALLOC;
+  }
+
+  // If we would have support for keyed topics, this could be KEEP_LAST and depth 1.
+  qos.history = RMW_QOS_POLICY_HISTORY_KEEP_ALL;
+  std::unique_ptr<rmw_subscription_t, std::function<void(rmw_subscription_t *)>>
+  subscription(
+    rmw_fastrtps_cpp::create_subscription(
+      participant_info.get(),
+      rosidl_typesupport_cpp::get_message_type_support_handle<ParticipantEntitiesInfo>(),
+      "_participant_info",
+      &qos,
+      &subscription_options,
+      false,  // our fastrtps typesupport doesn't support keyed topics
+      true),
+    [&](rmw_subscription_t * sub) {
+      if (RMW_RET_OK != rmw_fastrtps_shared_cpp::destroy_subscription(
+        eprosima_fastrtps_identifier,
+        participant_info.get(),
+        sub))
+      {
+        fprintf(
+          stderr, "Failed to destroy subscription after function: '"
+          RCUTILS_STRINGIFY(__function__) "' failed.\n");
+      }
+    });
+  if (!subscription) {
+    return RMW_RET_BAD_ALLOC;
+  }
+
+  common_context->gid = rmw_fastrtps_shared_cpp::create_rmw_gid(
+    eprosima_fastrtps_identifier, participant_info->participant->getGuid());
+  common_context->pub = publisher.get();
+  common_context->sub = subscription.get();
+
+  context->impl->common = common_context.get();
+  context->impl->participant_info = participant_info.get();
+
+  if (RMW_RET_OK != rmw_fastrtps_cpp::run_listener_thread(context)) {
+    return RMW_RET_ERROR;
+  }
+  common_context->graph_cache.add_participant(common_context->gid);
+
+  publisher.release();
+  subscription.release();
+  common_context.release();
+  participant_info.release();
+  return RMW_RET_OK;
+}
+
+rmw_ret_t
+rmw_fastrtps_cpp::register_node(rmw_context_t * context)
+{
+  assert(context);
+  assert(context->impl);
+
+  std::lock_guard<std::mutex> guard(context->impl->mutex);
+
+  if (!context->impl->count) {
+    rmw_ret_t ret = init_context_impl(context);
+    if (RMW_RET_OK != ret) {
+      return ret;
+    }
+  }
+  context->impl->count++;
+  return RMW_RET_OK;
+}
+
+rmw_ret_t
+rmw_fastrtps_cpp::unregister_node(rmw_context_t * context)
+{
+  assert(context);
+  assert(context->impl);
+
+  std::lock_guard<std::mutex> guard(context->impl->mutex);
+  if (0u == --context->impl->count) {
+    rmw_ret_t ret;
+    ret = rmw_fastrtps_cpp::join_listener_thread(context);
+    if (RMW_RET_OK != ret) {
+      return ret;
+    }
+
+    auto common_context = static_cast<rmw_dds_common::Context *>(context->impl->common);
+    auto participant_info = static_cast<CustomParticipantInfo *>(context->impl->participant_info);
+
+    common_context->graph_cache.remove_participant(common_context->gid);
```

Comment:
Can this throw or does it return something? If it does throw, is that ok?

Reply:
It has a return value that indicates if a change took place or not.
I'm unsure if logging a warning when false is desired, but I guess it might make sense.

It doesn't explicitly `throw` an exception, though it doesn't catch exceptions that may be thrown from the stdlib. I don't think that's ok, as throwing an exception throw a C boundary is UB.
Do you think that `GraphCache` itself should have all `nothrow` methods, or the caller should be responsable of doing a `try ... catch`?

----

Unrelated, but I noted some silmilar problems in the pre-existing codebase:
- There are some "potentially throwing" places without a try catch block. e.g.: [here](https://github.com/ros2/rmw_fastrtps/blob/bd5507b61c7285d6083e514ccd49f00b2945baeb/rmw_fastrtps_shared_cpp/src/rmw_service_names_and_types.cpp#L68) an `std::map` is used and all methods call don't have a try/catch block.
- All `std` containers aren't using an allocator. e.g.: [here](https://github.com/ros2/rmw_fastrtps/blob/bd5507b61c7285d6083e514ccd49f00b2945baeb/rmw_fastrtps_shared_cpp/include/rmw_fastrtps_shared_cpp/custom_participant_info.hpp#L192-L197) (and that one can cause an allocation anytime during the execution).
- Though using `goto` to handle errors in `c` is fine, in `cpp` that might be problematic. If an exception is thrown, the cleanup code after the `fail` label will never be executed. e.g.: [here](https://github.com/ros2/rmw_fastrtps/blob/bd5507b61c7285d6083e514ccd49f00b2945baeb/rmw_fastrtps_cpp/src/rmw_subscription.cpp#L146), where [a later call](https://github.com/ros2/rmw_fastrtps/blob/bd5507b61c7285d6083e514ccd49f00b2945baeb/rmw_fastrtps_cpp/src/rmw_subscription.cpp#L171) might potentially throw.
- There are some `memcpy` usage, where a plain `c` array is copied into an object, and doing the opposite too. See https://github.com/ros2/rmw_fastrtps/pull/346.

If you agree with some of those problems, I will create a ticket to iterate on this later.

Reply:
> I'm unsure if logging a warning when false is desired, but I guess it might make sense.

I would never advise suppressing an error case, but if you do, you should have a comment explaining why. Silently ignoring when something can fail is always bad, imo.

----

> Do you think that `GraphCache` itself should have all `nothrow` methods, or the caller should be responsible of doing a `try ... catch`?

Uh, that depends. Both can be correct to do, it just depends on what produces the least boilerplate and/or which is easiest to understand. I don't have specific guidance on this case.

-----

> There are some "potentially throwing" places without a try catch block. e.g.: here an std::map is used and all methods call don't have a try/catch block.

Yeah, that's unfortunate, but as we didn't write this originally there are lots of issues like that... If you compare it to the Connext rmw implementation, by comparison, we're very careful about exceptions, including try-catching constructors and using macros like this:

https://github.com/ros2/rmw/blob/35ae2ff54317e8e6f30acace2fbaccd57fb2c566/rmw/include/rmw/impl/cpp/macros.hpp#L29

But that doesn't mean we shouldn't try to be careful with new code.

> All std containers aren't using an allocator. e.g.: here (and that one can cause an allocation anytime during the execution).

Yup, some work was done in this vein and that's why I asked about the replacing of the `string_255` with `std::string` in some of your changes. It seemed like a backslide to me. But as with the exceptions, we should be careful with new code but not necessarily stop to fix everything you see. You could try to collect them into a list with links, which might help someone down the line.

> Though using `goto` to handle errors in c is fine, in cpp that might be problematic. If an exception is thrown, the cleanup code after the fail label will never be executed. e.g.: here, where a later call might potentially throw.

Sure, but if the `goto` is use in the implementation of an `rmw_*` function then throwing isn't allowed anyways.

But if its a different function, then yeah you could use the scope exit pattern instead.

> There are some memcpy usage, where a plain c array is copied into an object, and doing the opposite too. See #346.

Again, bad, but really doesn't change that we should be careful with error handling here.

> If you agree with some of those problems, I will create a ticket to iterate on this later.

Yeah, they all seem like good things to at least document for later.

Reply:
This one still needs resolving too. (check return code or change error handling I think)

Reply:
> I would never advise suppressing an error case, but if you do, you should have a comment explaining why. Silently ignoring when something can fail is always bad, imo.

Updated.

> Uh, that depends. Both can be correct to do, it just depends on what produces the least boilerplate and/or which is easiest to understand. I don't have specific guidance on this case.

I will make `GraphCache` methods `noexcept`, to avoid adding `try...catch` in every call in every rmw implementation that uses it.